pygments/lexers/jvm.py

```diff
@@ -156,226 +156,377 @@ class ScalaLexer(RegexLexer):
+            # the correct keyword. Note that soft modifiers can be followed by a
+            # sequence of regular modifiers; [a-z\s]* skips those, and we just
+            # check that the soft modifier is applied to a supported statement.
+            (r'\b(transparent|opaque|infix|open|inline)\b(?=[a-z\s]*\b'
+             r'(def|val|var|given|type|class|trait|object|enum)\b)', Keyword),
+        ],
+        'annotations': [
+            (u'@%s' % idrest, Name.Decorator),
+        ],
+        'using': [
+            # using is a soft keyword, can only be used in the first position of
+            # a parameter or argument list.
+            (r'(\()(\s*)(using)(\s)', bygroups(Punctuation, Text, Keyword, Text)),
+        ],
+        'declarations': [
+            (u'\\b(def)\\b(\\s*)%s(%s)?' % (notStartOfComment, anyId),
+             bygroups(Keyword, Text, Name.Function)),
+            (u'\\b(trait)\\b(\\s*)%s(%s)?' % (notStartOfComment, anyId),
+                bygroups(Keyword, Text, Name.Class)),
+            (u'\\b(?:(case)(\\s+))?(class|object|enum)\\b(\\s*)%s(%s)?' %
+                (notStartOfComment, anyId),
+                bygroups(Keyword, Text, Keyword, Text, Name.Class)),
+            (u'(?<!\\.)\\b(type)\\b(\\s*)%s(%s)?' % (notStartOfComment, anyId),
+                bygroups(Keyword, Text, Name.Class)),
+            (u'\\b(val|var)\\b', Keyword.Declaration),
+            (u'\\b(package)(\\s+)(object)\\b(\\s*)%s(%s)?' %
+                (notStartOfComment, anyId),
+                bygroups(Keyword, Text, Keyword, Text, Name.Namespace)),
+            (u'\\b(package)(\\s+)', bygroups(Keyword, Text), 'package'),
+            (u'\\b(given)\\b(\\s*)(%s)' % idUpper,
+                bygroups(Keyword, Text, Name.Class)),
+            (u'\\b(given)\\b(\\s*)(%s)?' % anyId, 
+                bygroups(Keyword, Text, Name)),
+        ],
+        'inheritance': [
+            (r'\b(extends|with|derives)\b(\s*)'
+             u'(%s|%s|(?=\\([^\\)]+=>)|(?=%s)|(?="))?' %
+                (idUpper, backQuotedId, plainid),
+                bygroups(Keyword, Text, Name.Class)),
+        ],
+        'extension': [
+            (r'\b(extension)(\s+)(?=[\[\(])', bygroups(Keyword, Text)),
+        ],
+        'end': [
+            # end is a soft keyword, should only be highlighted in certain cases
+            (r'\b(end)(\s+)(if|while|for|match|new|extension|val|var)\b',
+                bygroups(Keyword, Text, Keyword)),
+            (u'\\b(end)(\\s+)(%s)%s' % (idUpper, endOfLineMaybeWithComment),
+                bygroups(Keyword, Text, Name.Class)),
+            (u'\\b(end)(\\s+)(%s|%s)?%s' %
+                (backQuotedId, plainid, endOfLineMaybeWithComment),
+                bygroups(Keyword, Text, Name.Namespace)),
+        ],
+        'punctuation': [
+            (r'[{}()\[\];,.]', Punctuation),
+            (r'(?<!:):(?!:)', Punctuation),  
+        ],
+        'keywords': [
```

Comment:
these should all use `words()` which auto-escapes and optimizes the regex. Same for operators.

Reply:
Done. I've also fixed the storage modifiers to use `words`